### DISCUSSION TITLE
Naming: wrappers / APIs / packages

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#720] [General]
This is basically the collection of information and conclusions from earlier discussions regarding the subject of names of entities, namespaces, packages, wrappers, etc. I collect this here to spare myself from searching for it again the next time a new wrapper is proposed (see #719).

For reference, please checkout https://github.com/zxing-cpp/zxing-cpp/pull/337#issuecomment-1136363104 and the followup discussion as well as https://github.com/zxing-cpp/zxing-cpp/issues/204#issuecomment-821893053. If it were not for the limitation of not allowing '-' in some contexts and historical reasons, I'd prefer if it was called *zxing-cpp* (most of the time). Even though having different names for the same thing is far from ideal, each of the following has its merits or it's reasoning. I'll do my very best to make the situation not worse.

**zxing-cpp:**
 * the main name nowadays as far as I'm concerned, mostly because of the discoverability of the github url
 * (part of) the package name for the Android (mavencentral), iOS (cocoapods.org), Python (pypi.org) and Rust (crates.io) wrappers
 * Android rootProject.name

**zxingcpp:**
 * used everywhere where a '-' is not allowed but a lower-case name is idiomatic
 * the android project name
 * the python module name
 * the rust crate name
 * email: zxingcpp at gmail dot com

**ZXingCpp:**
 * the C# namespace in the .NET wrapper
 * the package name of the NuGet .NET package, since there are firm guidelines of how to name a package
 * the SwiftPM name and the module name to be imported

**ZXing:**
 * the C++ namespace where the core library code is found, since it is a C++ namespace there was no reason to call it ZXingCpp
 * the name of the c++ library binary, still here merely for historical and backward compatibility reasons
 * the prefix for all entities in the C-API
 * the include directory name of the c/c++ header files
 * the cmake project and library target (`ZXing::ZXing`)

**ZXing-C++:**
 * the name used in documentation/README contexts only, this could go away rather easily

**zxing:**
 * the name of WASM module (this should change)

**Comment by duanqn** [comment]
Add one more --

### nu-book-zxing-cpp:

* The package name used in vcpkg.

I have no idea why it got this name but I recognize that it would be hard to change it.

**Reply by axxel** [reply]
This is not in the same category as to the other mentioned names because it is outside the control of this project/repository. The same goes for all the linux distribution packages, the mingw and conan package. Although it might be worth pointing out that most, if not all of them use the term *zxing-cpp*.

The *nu-book* part of this name comes from the original github repository owner but I don't know why it was originally used in the package name. Maybe there was another one called *zxing-cpp* already? There does not seem to be one now. Maybe they would accept a PR that adds a new vcpkg with that name (and updates to 2.2.1 at the same time). The old one could then simply stay there?

**Reply by duanqn** [reply]
I actually tried to submit a name-changing PR but they did not want such changes. Looks like the name will not be changed in the near future.

**Reply by axxel** [reply]
Thanks for trying. I was looking for that PR/Issue on their github page but could not find it. ?

Unrelated, but could you maybe send them a PR to upgrade to 2.2.1? Seems like no one is picking this up.

**Reply by duanqn** [reply]
This is the discussion about naming: https://github.com/microsoft/vcpkg/issues/27704#issuecomment-1308243372
Sure. I will submit an issue about upgrading to 2.2.1

**Comment by axxel** [comment]
In the c-API `zxing_` has now been replaced by `ZXing_`. Luckily this has not been released, yet. :)

**Reply by pranavtriguna** [reply]
Just to clarify, this means the c-API wrapper, correct? There's no C-API as of right now, am I right ?
Thanks!

**Reply by axxel** [reply]
There is a `BUILD_C_API` cmake build option. If you set this to `ON` then the c-api will be build-in to the c++ library. This was added after the most recent 2.2.1 release. But it will be part of the next release. I will then also set the value to `ON` by default, since this feature is virtually free in terms of binary size and I'd like Linux distributions to pick it up automatically.

**Reply by pranavtriguna** [reply]
Awesome, looking forward to it !
Thanks!

**Reply by axxel** [reply]
Update: the cmake option is now called `ZXING_C_API` and is going to `ON` by default in the upcoming 2.3.0 release.
